### PR TITLE
scalafix: 0.10.0 -> 0.12.0

### PR DESCRIPTION
**Pull request by hakan-demirli**
## Description of changes
Bumps scalafix. Supersedes #275815

It also changes the `let ... in ...` syntax.
Nixpkgs often falls behind due to sheer number of PRs and limited number of maintainers. I frequently override src and hash to workaround the issue temporarly. It's not possible to override the variables defined in `let ... in ...`.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/tools/scalafix/default.nix [hunk 1]
```diff
@@ -1,47 +1,49 @@
...
+{
+  lib,
+  stdenv,
+  jre,
+  coursier,
+  makeWrapper,
+  installShellFiles,
+  setJavaClassPath,
+}:
+stdenv.mkDerivation rec {
```
Comment (liff) [inline pkgs/development/tools/scalafix/default.nix:10]: An alternative here, to make overriding slightly easier, is to use [the function variant](https://nixos.org/manual/nixpkgs/stable/#mkderivation-recursive-attributes):

```suggestion
stdenv.mkDerivation (finalAttrs: {
```

If you do this you also need to change `pname` to `finalAttrs.pname` below.

File: pkgs/development/tools/scalafix/default.nix [hunk 2]
```diff
@@ -1,47 +1,49 @@
...
   '';
 
   installCheckPhase = ''
-    $out/bin/${baseName} --version | grep -q "${version}"
+    $out/bin/${pname} --version | grep -q "${version}"
   '';
 
   meta = with lib; {
     description = "Refactoring and linting tool for Scala";
-    mainProgram = "scalafix";
```
Comment (liff) [inline pkgs/development/tools/scalafix/default.nix:42]: I don’t think this should be removed.

**Review by liff — CHANGES_REQUESTED** [review]
Commit message should match the pull request title.

Apart from that: tested and works fine.

**Comment by hakan-demirli** [comment]
Thanks for the feedback. I will be more careful next time.

**Review by liff — APPROVED** [review]

**Review by Tomahna — APPROVED** [review]
Tested on nixos.

File: pkgs/development/tools/scalafix/default.nix [hunk 3]
```diff
@@ -1,47 +1,52 @@
...
+      installShellCompletion --cmd ${finalAttrs.pname} \
+        --bash <($out/bin/${finalAttrs.pname} --bash) \
+        --zsh  <($out/bin/${finalAttrs.pname} --zsh)
+    '';
 
-  installCheckPhase = ''
-    $out/bin/${baseName} --version | grep -q "${version}"
-  '';
+    installCheckPhase = ''
+      $out/bin/${finalAttrs.pname} --version | grep -q "${finalAttrs.version}"
```
Comment (Aleksanaa) [inline pkgs/development/tools/scalafix/default.nix:41]: This logic can be moved into `passthru.tests` with `testers.testVersion`.

File: pkgs/development/tools/scalafix/default.nix [hunk 4]
```diff
@@ -1,47 +1,52 @@
...
-    license = licenses.bsd3;
-    maintainers = [ maintainers.tomahna ];
-  };
-}
+    meta = with lib; {
+      description = "Refactoring and linting tool for Scala";
+      mainProgram = "scalafix";
+      homepage = "https://scalacenter.github.io/scalafix/";
+      license = licenses.bsd3;
+      maintainers = [maintainers.tomahna];
```
Comment (Aleksanaa) [inline pkgs/development/tools/scalafix/default.nix:49]:
```suggestion
      maintainers = [maintainers.tomahna];
      sourceProvenance = with sourceTypes; [ binaryBytecode ];
```

**Review by Aleksanaa — APPROVED** [review]

**Pull request merged**